api.planx.uk/modules/auth/service.ts

```diff
@@ -1,34 +1,49 @@
 import { sign } from "jsonwebtoken";
-import { adminGraphQLClient as adminClient } from "../../hasura";
-import { gql } from "graphql-request";
+import { $admin } from "../../client";
+import { User, Role } from "@opensystemslab/planx-core/types";
 
-export const buildJWT = async (email: string | undefined) => {
-  const { users } = await adminClient.request(
-    gql`
-      query ($email: String!) {
-        users(where: { email: { _eq: $email } }, limit: 1) {
-          id
-        }
-      }
-    `,
-    { email },
-  );
-
-  if (!users.length) return;
-
-  const { id } = users[0];
-
-  const hasura = {
-    "x-hasura-allowed-roles": ["platformAdmin"],
-    "x-hasura-default-role": "platformAdmin",
-    "x-hasura-user-id": id.toString(),
-  };
+export const buildJWT = async (email: string): Promise<string | undefined> => {
+  const user = await $admin.user.getByEmail(email);
+  if (!user) return;
 
   const data = {
-    sub: id.toString(),
-    "https://hasura.io/jwt/claims": hasura,
+    sub: user.id.toString(),
+    "https://hasura.io/jwt/claims": generateHasuraClaimsForUser(user),
   };
 
   const jwt = sign(data, process.env.JWT_SECRET!);
   return jwt;
 };
+
+const generateHasuraClaimsForUser = (user: User) => ({
+  "x-hasura-allowed-roles": getAllowedRolesForUser(user),
+  "x-hasura-default-role": getDefaultRoleForUser(user),
+  "x-hasura-user-id": user.id.toString(),
+});
+
+/**
+ * Get all possible roles for this user
+ * Requests made outside this scope will not be authorised by Hasura
+ */
+const getAllowedRolesForUser = (user: User): Role[] => {
+  const teamRoles = user.teams.map((teamRole) => teamRole.role);
+  const allowedRoles: Role[] = [
+    "public", // Allow public access
+    "teamEditor", // Least privileged role for authenticated users - required for Editor access
+    ...teamRoles, // User specific roles
+  ];
```

Comment:
is `teamViewer` intentionally omitted here? Not sure I totally understand why

Reply:
This is a good question and something I don't think I've spelt out anywhere really.

`teamViewer` is intentionally left out here as it's not currently backed up by a Hasura role. We could do with talking about how we handle `teamViewer` and what this actually means.

 - `teamEditor` grants read-only permissions to all flows/operations/settings etc, but only insert, update, and delete when you're a member of that team
 - This pretty much describes what a `teamViewer` should be? Read only access to all teams - we don't need a relationship to a specific team. Having a record for every user and every team in `team_members` wouldn't be necessary.
 
 So here's my proposal - 
  - Drop `teamViewer` from the `user_roles` enum table - it doesn't really make sense as it doesn't require a relationship to a team
  - Make a `readOnly` role for people like DLUHC users
  - This would be the new default for logged in users